contrib/gogearbox/gearbox.v1/gearbox_test.go

```diff
@@ -0,0 +1,305 @@
+		fcc := &gearboxutil.FasthttpCarrier{ReqHeader: &gb.Context().Request.Header}
+
+		pspan := tracer.StartSpan("test")
+		err := tracer.Inject(pspan.Context(), fcc)
+		if err != nil {
+			t.Fatalf("Trace injection failed")
+		}
+		Middleware(WithServiceName("gb"))(gb)
+		sctx, err := tracer.Extract(fcc)
+		if err != nil {
+			t.Fatalf("Trace extraction failed")
+		}
+		assert.Equal(sctx.TraceID(), pspan.Context().TraceID())
+		assert.Equal(sctx.SpanID(), pspan.Context().SpanID())
+	})
+	t.Run("req-context", func(t *testing.T) {
+		assert := assert.New(t)
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		reqctx := newReqCtx(200)
+		gb := &GearboxContextMock{requestCtx: reqctx}
+		fcc := &gearboxutil.FasthttpCarrier{ReqHeader: &gb.Context().Request.Header}
+
+		pspan := tracer.StartSpan("test")
+		err := tracer.Inject(pspan.Context(), fcc)
+		if err != nil {
+			t.Fatalf("Trace injection failed")
+		}
+		Middleware(WithServiceName("gb"))(gb)
+		span, ok := tracer.SpanFromContext(gb.Context())
+		assert.True(ok)
+		assert.Equal(span.(mocktracer.Span).TraceID(), pspan.(mocktracer.Span).TraceID())
+		assert.Equal(span.(mocktracer.Span).ParentID(), pspan.(mocktracer.Span).SpanID())
+	})
+}
+func BenchmarkGearboxMiddleware(b *testing.B) {
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	reqctx := newReqCtx(200)
+	gb := &GearboxContextMock{requestCtx: reqctx}
+	for i := 0; i < b.N; i++ {
+		Middleware()(gb)
+	}
+}
+func BenchmarkGearboxMiddlewareWithOptions(b *testing.B) {
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	reqctx := newReqCtx(200)
+	gb := &GearboxContextMock{requestCtx: reqctx}
+	for i := 0; i < b.N; i++ {
+		Middleware(WithServiceName("gb"), WithStatusCheck(customErrChecker), WithResourceNamer(resourceNamer), WithIgnoreRequest(ignoreResources))(gb)
+	}
+}
+
+// BenchmarkGearbox is intended to serve as a comparison between gearbox with trace middleware v other middleware.
+// MTOFF: Not sure if this is the right approach especially since the other benchmarks use GearboxContextMock
+func BenchmarkGearbox(b *testing.B) {
+	gb := gearbox.New()
+	logMiddleware := func(ctx gearbox.Context) {
+		fmt.Println("log message!")
+		ctx.Next()
+	}
+	gb.Use(logMiddleware)
+}
+
+func customErrChecker(statusCode int) bool {
+	return statusCode >= 600
+}
+
+var customRsc = "custom resource"
+
+func resourceNamer(gctx gearbox.Context) string {
+	return customRsc
+}
+func ignoreResources(c gearbox.Context) bool {
+	return strings.HasPrefix(string(c.Context().URI().Path()), "/any")
+}
+
+// GearboxContextMock provides a mock implementation the Gearbox library to be used in testing.
+// It fulfills the Gearbox.Context interface and its methods mock some of the core behavior of
+// Gearbox.Context when handling real HTTP traffic, without needing to spin up a server.
+// See: https://pkg.go.dev/github.com/gogearbox/gearbox#Context
+type GearboxContextMock struct {
+	requestCtx *fasthttp.RequestCtx
+	index      int
+}
+
```

Comment:
Gearbox doesn't have any testing utils nor any implementation of some kind of in-memory listener. So I couldn't spin up a client & server. Therefore I chose to mock Gearbox.Context.

Only some of the methods required to fulfill the interface were necessary for my tests, so other methods have a dummy implementation. I've never implemented mock-testing so pls lmk if this should get cleaned up & how!